File: ShaliniKumari_BanasthaliVidyapith/Arrays/Problem7.cpp
#include <iostream>
#include <vector>
using namespace std;
int maxSubArray(vector<int>& nums) {
    int max_=nums[0];
    int curr_=nums[0];
    for(int i=1;i<nums.size();i++){
        curr_=max(nums[i],curr_+nums[i]);
        max_=max(curr_,max_);
    }
    return max_;
}
int main() {
    vector<int> nums= {-2,1,-3,4,-1,2,1,-5,4};
    int answer = maxSubArray(nums);
    cout << answer << endl;
    return 0;
}

File: Arrays_and_Strings/Problem7.java
/*

Problem 7:


Given an integer array nums, find the subarray with the largest sum, and return its sum.


Example 1:
Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.


Example 2:
Input: nums = [1]
Output: 1
Explanation: The subarray [1] has the largest sum 1.


Example 3:
Input: nums = [5,4,-1,7,8]
Output: 23
Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.


Constraints:
1 <= nums.length <= 105
-104 <= nums[i] <= 104
 */


import java.util.*;
public class Problem7 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] nums = new int[n];
        for(int i=0; i<n; i++){
            nums[i] = sc.nextInt();
        }
        System.out.println(maxSubArray(nums));
        sc.close();
    }
    public static int maxSubArray(int[] nums) {
        int maxEndingHere = nums[0];
        int maxSoFar = nums[0];
        for (int i = 1; i < nums.length; i++) {
            maxEndingHere = Math.max(nums[i], nums[i] + maxEndingHere);
            maxSoFar = Math.max(maxSoFar, maxEndingHere);
        }
        return maxSoFar;
    }    
}
